File: rust/crates/corlinman-gateway/src/middleware/admin_auth.rs
//! Auth guard for `/admin/*`.
//!
//! Two credentials get you past the guard, checked in order:
//!   1. `Cookie: corlinman_session=<token>` validated against
//!      [`super::admin_session::AdminSessionStore`] (S5 T1) — the normal
//!      UI path after `/admin/login`.
//!   2. `Authorization: Basic base64(user:pass)` verified against
//!      `config.admin.username` + `password_hash` (argon2id) — kept as a
//!      fallback for curl / CI / the initial login request itself.
//!
//! Both paths short-circuit the other: a cookie hit skips Basic entirely,
//! and missing/expired cookie falls through to Basic rather than 401. This
//! keeps the old M6 contract (HTTP Basic works) intact while letting the
//! UI authenticate once and rely on the cookie from then on.
//!
//! `AdminAuthState` clones cheaply (all fields are `Arc`), so rotating the
//! admin password or invalidating a session takes effect on the next
//! request without restart.

use std::sync::Arc;

use arc_swap::ArcSwap;
use argon2::{password_hash::PasswordHash, Argon2, PasswordVerifier};
use axum::{
    body::Body,
    extract::State,
    http::{header, Request, StatusCode},
    middleware::Next,
    response::{IntoResponse, Response},
    Json,
};
use base64::Engine;
use corlinman_core::config::Config;
use serde_json::json;

use super::admin_session::AdminSessionStore;

/// Cookie name carrying the opaque session token issued by `/admin/login`.
/// Exported so the login/logout handlers can write exactly the same name.
pub const SESSION_COOKIE_NAME: &str = "corlinman_session";

/// Cloneable bundle of state that the admin auth middleware + admin handlers
/// need. `config` is shared with the rest of the gateway; verifying per
/// request lets admins rotate credentials without a restart. `session_store`
/// is `None` in legacy / test setups that only want Basic auth — the
/// middleware then falls straight through to the Basic path.
#[derive(Clone)]
pub struct AdminAuthState {
    pub config: Arc<ArcSwap<Config>>,
    pub session_store: Option<Arc<AdminSessionStore>>,
}

impl AdminAuthState {
    pub fn new(config: Arc<ArcSwap<Config>>) -> Self {
        Self {
            config,
            session_store: None,
        }
    }

    /// Fluent: attach a session store so cookie auth is accepted.
    pub fn with_session_store(mut self, store: Arc<AdminSessionStore>) -> Self {
        self.session_store = Some(store);
        self
    }
}

/// Response payload for a 401. Matches the shape used by the rest of the
/// gateway so the UI's `CorlinmanApiError` parser can display a useful
/// message.
fn unauthorized(reason: &'static str) -> Response {
    (
        StatusCode::UNAUTHORIZED,
        [(header::WWW_AUTHENTICATE, r#"Basic realm="corlinman-admin""#)],
        Json(json!({
            "error": "unauthorized",
            "reason": reason,
        })),
    )
        .into_response()
}

/// Parse `Authorization: Basic <base64>` → `(user, pass)`. Returns `None`
/// for any malformed or non-Basic header.
fn parse_basic(header_value: &str) -> Option<(String, String)> {
    let rest = header_value.strip_prefix("Basic ")?.trim();
    let decoded = base64::engine::general_purpose::STANDARD
        .decode(rest)
        .ok()?;
    let s = String::from_utf8(decoded).ok()?;
    let (user, pass) = s.split_once(':')?;
    Some((user.to_string(), pass.to_string()))
}

/// Extract a named cookie value from a `Cookie:` header. Scan manually so we
/// don't pull a cookie parser crate for two fields. Returns `None` if the
/// header is absent or the cookie isn't present.
pub(crate) fn extract_cookie(header_value: &str, name: &str) -> Option<String> {
    for part in header_value.split(';') {
        let part = part.trim();
        if let Some((k, v)) = part.split_once('=') {
            if k.trim() == name {
                return Some(v.trim().to_string());
            }
        }
    }
    None
}

/// Verify `password` against an argon2id hash string. Any parse / verify
/// failure yields `false`; we never distinguish "wrong password" from
/// "malformed stored hash" in the response to avoid leaking hash shape.
pub(crate) fn argon2_verify(password: &str, stored_hash: &str) -> bool {
    let Ok(parsed) = PasswordHash::new(stored_hash) else {
        tracing::warn!("admin.password_hash is not a valid argon2 PHC string");
        return false;
    };
    Argon2::default()
        .verify_password(password.as_bytes(), &parsed)
        .is_ok()
}

/// Axum middleware function. Attach via
/// `Router::layer(from_fn_with_state(state, require_admin))`.
pub async fn require_admin(
    State(state): State<AdminAuthState>,
    req: Request<Body>,
    next: Next,
) -> Response {
    let cfg = state.config.load();
    let Some(expected_user) = cfg.admin.username.as_deref() else {
        // No admin credentials configured at all — fail closed.
        return unauthorized("admin_not_configured");
    };
    let Some(expected_hash) = cfg.admin.password_hash.as_deref() else {
        return unauthorized("admin_not_configured");
    };

    // 1) Cookie path — only if a session store is wired.
    if let Some(store) = state.session_store.as_ref() {
        if let Some(cookie_header) = req
            .headers()
            .get(header::COOKIE)
            .and_then(|v| v.to_str().ok())
        {
            if let Some(token) = extract_cookie(cookie_header, SESSION_COOKIE_NAME) {
                if store.validate(&token).is_some() {
                    return next.run(req).await;
                }
                // Cookie present but invalid/expired: fall through to
                // Basic auth rather than 401 so curl/CI still works.
            }
        }
    }

    // 2) Basic auth fallback.
    let Some(auth_header) = req
        .headers()
        .get(header::AUTHORIZATION)
        .and_then(|v| v.to_str().ok())
    else {
        return unauthorized("missing_authorization");
    };

    let Some((user, pass)) = parse_basic(auth_header) else {
        return unauthorized("malformed_authorization");
    };

    if user != expected_user {
        return unauthorized("invalid_credentials");
    }
    if !argon2_verify(&pass, expected_hash) {
        return unauthorized("invalid_credentials");
    }

    next.run(req).await
}

#[cfg(test)]
mod tests {
    use super::*;
    use arc_swap::ArcSwap;
    use argon2::password_hash::{PasswordHasher, SaltString};
    use axum::body::Body;
    use axum::http::Request;
    use axum::routing::get;
    use axum::Router;
    use std::sync::Arc;
    use std::time::Duration as StdDuration;
    use tower::ServiceExt;

    fn hash_password(password: &str) -> String {
        let salt = SaltString::encode_b64(b"corlinman_test_salt_bytes_16").unwrap();
        Argon2::default()
            .hash_password(password.as_bytes(), &salt)
            .unwrap()
            .to_string()
    }

    fn state_with(user: Option<&str>, password: Option<&str>) -> AdminAuthState {
        let mut cfg = Config::default();
        cfg.admin.username = user.map(str::to_string);
        cfg.admin.password_hash = password.map(hash_password);
        AdminAuthState::new(Arc::new(ArcSwap::from_pointee(cfg)))
    }

    fn state_with_session(
        user: Option<&str>,
        password: Option<&str>,
        store: Arc<AdminSessionStore>,
    ) -> AdminAuthState {
        state_with(user, password).with_session_store(store)
    }

    fn app(state: AdminAuthState) -> Router {
        Router::new()
            .route("/ping", get(|| async { "pong" }))
            .layer(axum::middleware::from_fn_with_state(state, require_admin))
    }

    fn basic_header(user: &str, pass: &str) -> String {
        let raw = format!("{user}:{pass}");
        format!(
            "Basic {}",
            base64::engine::general_purpose::STANDARD.encode(raw)
        )
    }

    #[tokio::test]
    async fn missing_header_is_401() {
        let app = app(state_with(Some("admin"), Some("secret")));
        let res = app
            .oneshot(Request::builder().uri("/ping").body(Body::empty()).unwrap())
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::UNAUTHORIZED);
        assert!(res.headers().contains_key(header::WWW_AUTHENTICATE));
    }

    #[tokio::test]
    async fn wrong_password_is_401() {
        let app = app(state_with(Some("admin"), Some("secret")));
        let res = app
            .oneshot(
                Request::builder()
                    .uri("/ping")
                    .header(header::AUTHORIZATION, basic_header("admin", "WRONG"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::UNAUTHORIZED);
    }

    #[tokio::test]
    async fn wrong_username_is_401() {
        let app = app(state_with(Some("admin"), Some("secret")));
        let res = app
            .oneshot(
                Request::builder()
                    .uri("/ping")
                    .header(header::AUTHORIZATION, basic_header("root", "secret"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::UNAUTHORIZED);
    }

    #[tokio::test]
    async fn correct_credentials_pass_through() {
        let app = app(state_with(Some("admin"), Some("secret")));
        let res = app
            .oneshot(
                Request::builder()
                    .uri("/ping")
                    .header(header::AUTHORIZATION, basic_header("admin", "secret"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::OK);
    }

    #[tokio::test]
    async fn no_admin_configured_is_401() {
        let app = app(state_with(None, None));
        let res = app
            .oneshot(
                Request::builder()
                    .uri("/ping")
                    .header(header::AUTHORIZATION, basic_header("admin", "secret"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::UNAUTHORIZED);
    }

    #[test]
    fn parse_basic_accepts_well_formed() {
        let h = basic_header("alice", "hunter2");
        let (u, p) = parse_basic(&h).unwrap();
        assert_eq!(u, "alice");
        assert_eq!(p, "hunter2");
    }

    #[test]
    fn parse_basic_rejects_non_basic() {
        assert!(parse_basic("Bearer xyz").is_none());
        assert!(parse_basic("Basic @@@not-base64@@@").is_none());
    }

    // --- cookie / session_store branch -----------------------------------

    #[test]
    fn extract_cookie_finds_named_value() {
        assert_eq!(
            extract_cookie("foo=bar; corlinman_session=abc123", SESSION_COOKIE_NAME),
            Some("abc123".into())
        );
        assert_eq!(
            extract_cookie("corlinman_session=xyz", SESSION_COOKIE_NAME),
            Some("xyz".into())
        );
        assert!(extract_cookie("foo=bar", SESSION_COOKIE_NAME).is_none());
    }

    #[tokio::test]
    async fn valid_cookie_lets_request_through() {
        let store = Arc::new(AdminSessionStore::new(StdDuration::from_secs(60)));
        let token = store.create("admin".into());
        let state = state_with_session(Some("admin"), Some("secret"), store);
        let res = app(state)
            .oneshot(
                Request::builder()
                    .uri("/ping")
                    .header(header::COOKIE, format!("{SESSION_COOKIE_NAME}={token}"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::OK);
    }

    #[tokio::test]
    async fn expired_cookie_falls_through_to_basic_auth() {
        // 0s TTL ⇒ token is already expired.
        let store = Arc::new(AdminSessionStore::new(StdDuration::from_secs(0)));
        let token = store.create("admin".into());
        tokio::time::sleep(StdDuration::from_millis(1100)).await;
        let state = state_with_session(Some("admin"), Some("secret"), store);

        // Expired cookie + valid Basic auth should succeed.
        let res = app(state.clone())
            .oneshot(
                Request::builder()
                    .uri("/ping")
                    .header(header::COOKIE, format!("{SESSION_COOKIE_NAME}={token}"))
                    .header(header::AUTHORIZATION, basic_header("admin", "secret"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::OK);

        // Expired cookie + no Basic auth → 401.
        let res = app(state)
            .oneshot(
                Request::builder()
                    .uri("/ping")
                    .header(header::COOKIE, format!("{SESSION_COOKIE_NAME}={token}"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::UNAUTHORIZED);
    }

    #[tokio::test]
    async fn cookie_takes_priority_over_basic_when_both_present() {
        let store = Arc::new(AdminSessionStore::new(StdDuration::from_secs(60)));
        let token = store.create("admin".into());
        let state = state_with_session(Some("admin"), Some("secret"), store);

        // Bogus Basic creds — the cookie should still let us through.
        let res = app(state)
            .oneshot(
                Request::builder()
                    .uri("/ping")
                    .header(header::COOKIE, format!("{SESSION_COOKIE_NAME}={token}"))
                    .header(header::AUTHORIZATION, basic_header("admin", "WRONG"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::OK);
    }

    #[tokio::test]
    async fn basic_auth_still_works_when_no_session_store() {
        // Mirrors the pre-S5 contract: even without a session store
        // wired, a valid Basic-auth request must pass.
        let state = state_with(Some("admin"), Some("secret"));
        let res = app(state)
            .oneshot(
                Request::builder()
                    .uri("/ping")
                    .header(header::AUTHORIZATION, basic_header("admin", "secret"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(res.status(), StatusCode::OK);
    }
}

File: python/packages/corlinman-server/src/corlinman_server/gateway/middleware/admin_auth.py
"""Auth guard for ``/admin/*``.

Python port of ``rust/crates/corlinman-gateway/src/middleware/admin_auth.rs``.

Two credentials get a request past the guard, checked in order:

1. ``Cookie: corlinman_session=<token>`` validated against
   :class:`~corlinman_server.gateway.middleware.admin_session.AdminSessionStore`
   — the normal UI path after ``/admin/login``.
2. ``Authorization: Basic base64(user:pass)`` verified via argon2id
   against an :class:`~corlinman_server.tenancy.AdminRow` looked up
   by ``(tenant_id, username)`` in :class:`~corlinman_server.tenancy.AdminDb`.
   This is the fallback that keeps curl / CI / the initial login request
   working when no cookie is yet established.

Both paths short-circuit the other: a cookie hit skips Basic entirely,
and a missing / expired cookie falls through to Basic instead of 401.
This mirrors the Rust impl byte-for-byte so the existing UI contract
holds.

On success the middleware stashes:

* ``request.state.admin_user`` — the username (str).
* ``request.state.admin_tenant`` — the :class:`TenantId` the credential
  is scoped to (defaults to ``TenantId.legacy_default()`` for Basic auth
  when no tenant is encoded in the request).
* ``request.state.admin_session`` — the resolved
  :class:`~corlinman_server.gateway.middleware.admin_session.AdminSession`
  when authentication came via cookie (``None`` on Basic-auth hits).

These keys are also what :func:`require_admin` (the ``Depends`` factory
sibling) returns, packaged into an :class:`AdminPrincipal`.
"""

from __future__ import annotations

import base64
from dataclasses import dataclass
from typing import Any

import structlog
from fastapi import Depends, HTTPException, Request, status
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse, Response
from starlette.types import ASGIApp

from corlinman_server.gateway.middleware.admin_session import (
    AdminSession,
    AdminSessionStore,
)
from corlinman_server.tenancy import AdminDb, TenantId, default_tenant

logger = structlog.get_logger(__name__)


#: Cookie name carrying the opaque session token issued by ``/admin/login``.
#: Exported so the login/logout handlers can write exactly the same name.
SESSION_COOKIE_NAME: str = "corlinman_session"


#: Path prefixes the middleware gates. Mirrors the Rust gateway's
#: ``/admin/*`` mount point. ``/admin/login`` is intentionally included
#: in the gated set so the login handler can also be reached by a
#: pre-existing cookie — the Rust router mounts ``/admin/login``
#: *outside* this layer; the Python port lets a route-level
#: ``Depends`` opt out per-endpoint instead (e.g. login itself).
DEFAULT_ADMIN_PREFIXES: tuple[str, ...] = ("/admin/",)


# ---------------------------------------------------------------------------
# Cloneable state
# ---------------------------------------------------------------------------


@dataclass
class AdminAuthState:
    """Bundle of handles the admin auth middleware reads on every request.

    ``admin_db`` looks up the :class:`AdminRow` and its argon2id hash;
    ``session_store`` is the in-memory cookie validator. Either can be
    ``None`` for tests / partial wiring — a missing ``admin_db`` fails
    closed (401 ``admin_not_configured``); a missing ``session_store``
    means cookies are never consulted and every request must present
    Basic auth.

    ``default_tenant_id`` is the tenant the middleware scopes Basic-auth
    lookups to when the request doesn't otherwise carry tenant context.
    Mirrors the Rust crate's reliance on ``config.admin.username`` /
    ``password_hash`` which are implicitly scoped to the legacy default.
    """

    admin_db: AdminDb | None = None
    session_store: AdminSessionStore | None = None
    default_tenant_id: TenantId = default_tenant()
    protected_prefixes: tuple[str, ...] = DEFAULT_ADMIN_PREFIXES


# ---------------------------------------------------------------------------
# Header / cookie parsing — pure helpers, exported for tests.
# ---------------------------------------------------------------------------


def parse_basic(header_value: str) -> tuple[str, str] | None:
    """Parse ``Authorization: Basic <base64>`` → ``(user, pass)``.

    Returns ``None`` for any malformed or non-Basic header. Byte-for-byte
    match with the Rust ``parse_basic`` helper.
    """

    if not header_value.lower().startswith("basic "):
        return None
    rest = header_value[6:].strip()
    try:
        decoded = base64.b64decode(rest, validate=True)
    except (ValueError, base64.binascii.Error):
        return None
    try:
        s = decoded.decode("utf-8")
    except UnicodeDecodeError:
        return None
    if ":" not in s:
        return None
    user, _, password = s.partition(":")
    return user, password


def extract_cookie(header_value: str, name: str) -> str | None:
    """Pull a named cookie value out of a ``Cookie:`` header.

    Hand-rolled scan instead of pulling a cookie parser dep — matches
    the Rust ``extract_cookie`` helper. Returns ``None`` if the header
    is absent or the cookie isn't present.
    """

    for part in header_value.split(";"):
        part = part.strip()
        if "=" not in part:
            continue
        k, _, v = part.partition("=")
        if k.strip() == name:
            return v.strip()
    return None


def argon2_verify(password: str, stored_hash: str) -> bool:
    """Verify ``password`` against an argon2id PHC-encoded hash.

    Any parse / verify failure yields ``False`` — we never distinguish
    "wrong password" from "malformed stored hash" so the response
    doesn't leak hash shape. Mirrors the Rust ``argon2_verify`` helper.
    """

    try:
        from argon2 import PasswordHasher
        from argon2.exceptions import (
            InvalidHashError,
            VerificationError,
            VerifyMismatchError,
        )
    except ImportError:
        logger.error("admin_auth.argon2_missing")
        return False

    hasher = PasswordHasher()
    try:
        return hasher.verify(stored_hash, password)
    except (VerifyMismatchError, VerificationError, InvalidHashError):
        return False
    except Exception:  # noqa: BLE001 — defensive: never leak hash shape
        logger.warning("admin_auth.argon2_verify_unexpected")
        return False


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------


def _unauthorized(reason: str) -> JSONResponse:
    """401 response with the canonical envelope shape."""

    return JSONResponse(
        {"error": "unauthorized", "reason": reason},
        status_code=status.HTTP_401_UNAUTHORIZED,
        headers={"WWW-Authenticate": 'Basic realm="corlinman-admin"'},
    )


def _resolve_state(request: Request) -> AdminAuthState | None:
    state = getattr(request.app.state, "admin_auth", None)
    if isinstance(state, AdminAuthState):
        return state
    return None


def _path_is_protected(path: str, prefixes: tuple[str, ...]) -> bool:
    return any(path.startswith(p) for p in prefixes)


@dataclass(frozen=True)
class _AuthResult:
    """Outcome of running the auth chain. Either ``ok`` is populated
    with the :class:`AdminPrincipal`, or ``reason`` carries the 401
    short-circuit code."""

    ok: "AdminPrincipal | None" = None
    reason: str | None = None


async def _run_auth_chain(
    request: Request, state: AdminAuthState
) -> _AuthResult:
    """Cookie-then-Basic auth chain shared by the middleware and the
    ``Depends`` sibling. Returns either a populated principal or the
    canonical 401 reason."""

    if state.admin_db is None:
        return _AuthResult(reason="admin_not_configured")

    # 1) Cookie path — only consulted when a session store is wired.
    if state.session_store is not None:
        cookie_header = request.headers.get("cookie")
        if cookie_header is not None:
            token = extract_cookie(cookie_header, SESSION_COOKIE_NAME)
            if token is not None:
                session = state.session_store.validate(token)
                if session is not None:
                    principal = AdminPrincipal(
                        user=session.user,
                        tenant=state.default_tenant_id,
                        session=session,
                    )
                    return _AuthResult(ok=principal)
                # Cookie present but invalid/expired: fall through to
                # Basic rather than 401 so curl / CI still works.

    # 2) Basic auth fallback.
    auth_header = request.headers.get("authorization")
    if auth_header is None:
        return _AuthResult(reason="missing_authorization")

    parsed = parse_basic(auth_header)
    if parsed is None:
        return _AuthResult(reason="malformed_authorization")

    user, password = parsed

    # Look up the admin row. The Rust port reads
    # ``config.admin.username`` + ``password_hash`` directly; the Python
    # port has moved that data into ``tenant_admins`` on the AdminDb so
    # we route through it. The lookup is scoped to the default tenant
    # unless something further up has annotated otherwise.
    admin_row = None
    try:
        admin_row = await _lookup_admin(state.admin_db, state.default_tenant_id, user)
    except Exception as exc:  # noqa: BLE001
        logger.warning(
            "admin_auth.lookup_failed",
            tenant=state.default_tenant_id.as_str(),
            user=user,
            error=str(exc),
        )
        return _AuthResult(reason="invalid_credentials")

    if admin_row is None:
        return _AuthResult(reason="invalid_credentials")

    if not argon2_verify(password, admin_row.password_hash):
        return _AuthResult(reason="invalid_credentials")

    return _AuthResult(
        ok=AdminPrincipal(
            user=user,
            tenant=state.default_tenant_id,
            session=None,
        )
    )


async def _lookup_admin(
    admin_db: AdminDb, tenant: TenantId, username: str
) -> Any | None:
    """Look up ``(tenant, username)`` in ``tenant_admins``. Tries
    :meth:`AdminDb.get_admin` first (the preferred CRUD path); falls
    back to scanning :meth:`AdminDb.list_admins` if the former is
    missing on older revisions.
    """

    getter = getattr(admin_db, "get_admin", None)
    if callable(getter):
        return await getter(tenant, username)
    lister = getattr(admin_db, "list_admins", None)
    if callable(lister):
        for row in await lister(tenant):
            if getattr(row, "username", None) == username:
                return row
    return None


# ---------------------------------------------------------------------------
# Principal — the thing handlers see.
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class AdminPrincipal:
    """Resolved admin identity. Stashed on ``request.state.admin_*`` and
    returned by :func:`require_admin`."""

    user: str
    tenant: TenantId
    session: AdminSession | None


# ---------------------------------------------------------------------------
# Middleware
# ---------------------------------------------------------------------------


class AdminAuthMiddleware(BaseHTTPMiddleware):
    """Cookie + HTTP Basic guard for ``/admin/*``.

    Construction takes an explicit :class:`AdminAuthState`; the same
    state is published on ``app.state.admin_auth`` so :func:`require_admin`
    and admin handlers can read it.
    """

    def __init__(
        self,
        app: ASGIApp,
        state: AdminAuthState | None = None,
    ) -> None:
        super().__init__(app)
        self._state = state or AdminAuthState()

    async def dispatch(
        self,
        request: Request,
        call_next: Any,
    ) -> Response:
        state = _resolve_state(request) or self._state

        if not _path_is_protected(request.url.path, state.protected_prefixes):
            return await call_next(request)

        result = await _run_auth_chain(request, state)
        if result.ok is None:
            assert result.reason is not None
            return _unauthorized(result.reason)

        principal = result.ok
        request.state.admin_user = principal.user
        request.state.admin_tenant = principal.tenant
        request.state.admin_session = principal.session

        return await call_next(request)


def install_admin_auth_middleware(
    app: Any,
    *,
    admin_db: AdminDb | None = None,
    session_store: AdminSessionStore | None = None,
    default_tenant_id: TenantId | None = None,
    protected_prefixes: tuple[str, ...] = DEFAULT_ADMIN_PREFIXES,
) -> AdminAuthState:
    """Attach :class:`AdminAuthMiddleware` to ``app``.

    Returns the :class:`AdminAuthState` instance so the caller can rebind
    handles after install (e.g. when ``admin_db`` is opened lazily). The
    same instance is published on ``app.state.admin_auth``.
    """

    state = AdminAuthState(
        admin_db=admin_db,
        session_store=session_store,
        default_tenant_id=default_tenant_id or default_tenant(),
        protected_prefixes=protected_prefixes,
    )
    app.state.admin_auth = state
    app.add_middleware(AdminAuthMiddleware, state=state)
    return state


# ---------------------------------------------------------------------------
# FastAPI ``Depends`` factory
# ---------------------------------------------------------------------------


def require_admin() -> Any:
    """Per-route guard returning the authenticated :class:`AdminPrincipal`.

    Usage::

        @router.get("/admin/something")
        async def handler(admin: AdminPrincipal = Depends(require_admin())):
            ...

    Reuses any principal the middleware already validated on this
    request (avoids a second DB round-trip / argon2 call). Raises
    :class:`HTTPException` 401 in the canonical envelope on failure.
    """

    async def dependency(request: Request) -> AdminPrincipal:
        # Reuse middleware-populated state when available.
        cached_user = getattr(request.state, "admin_user", None)
        if isinstance(cached_user, str):
            return AdminPrincipal(
                user=cached_user,
                tenant=getattr(request.state, "admin_tenant", default_tenant()),
                session=getattr(request.state, "admin_session", None),
            )

        state = _resolve_state(request)
        if state is None:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail={"error": "unauthorized", "reason": "admin_not_configured"},
                headers={"WWW-Authenticate": 'Basic realm="corlinman-admin"'},
            )

        result = await _run_auth_chain(request, state)
        if result.ok is None:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail={"error": "unauthorized", "reason": result.reason or "unknown"},
                headers={"WWW-Authenticate": 'Basic realm="corlinman-admin"'},
            )

        principal = result.ok
        request.state.admin_user = principal.user
        request.state.admin_tenant = principal.tenant
        request.state.admin_session = principal.session
        return principal

    return Depends(dependency)


__all__ = [
    "DEFAULT_ADMIN_PREFIXES",
    "SESSION_COOKIE_NAME",
    "AdminAuthMiddleware",
    "AdminAuthState",
    "AdminPrincipal",
    "argon2_verify",
    "extract_cookie",
    "install_admin_auth_middleware",
    "parse_basic",
    "require_admin",
]
